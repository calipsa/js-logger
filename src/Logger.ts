import os from 'os'
import { pid } from 'process'
import { format } from 'util'

import { isPlainObject } from 'lodash'

import levels from './levels'

const hostname = os.hostname()

/**
 * mutates `data`
 */
function extractErrors(data: any[]) {
  const len = data.length
  const errs: Error[] = []
  for (let i = 0; i < len; ++i) {
    const d = data[i]
    if (d instanceof Error) {
      errs.push(d)
      // eslint-disable-next-line no-param-reassign
      data[i] = d.message
    }
  }
  return errs
}

type Level = keyof typeof levels

type LevelNum = (typeof levels)[Level]

interface WithMsg {
  msg: any,
}

interface WithErr {
  err?: any,
}

export interface Log extends WithMsg, WithErr {
  name: string,
  hostname: string,
  pid: number,
  level: LevelNum,
  time: Date,
  // [key: string]: any,
}

interface Serializers {
  [prop: string]: (o: {}) => {},
}

export interface Options {
  name: string,
  minLevel?: Level,
  extraProps?: {},
  serializers?: Serializers,
  transform?: (o: Log) => any,
}

type BaseLogger = Pick<Console, Level>

export default class Logger implements BaseLogger {
  private readonly options: Options
  private readonly name: string
  private readonly minLevel: LevelNum
  private readonly extraProps: {}
  private readonly serializers: Serializers
  private readonly transform: (o: any) => any

  constructor(options: Options) {
    this.options = options
    this.name = options.name
    this.minLevel = levels[options.minLevel ?? 'info']
    this.extraProps = options.extraProps ?? {}
    this.serializers = options.serializers ?? {}
    this.transform = options.transform ?? JSON.stringify
  }

  private serialize(o: {}) {
    return Object.entries(this.serializers).reduce(
      (prev, [prop, transform]) => prop in o
        ? ({
          ...prev,
          // @ts-ignore
          [prop]: transform(o[prop]),
        })
        : prev,
      {
        ...o,
      },
    )
  }

  private write(lvl: Level, ...data: any[]) {
    if (levels[lvl] < this.minLevel) {
      return
    }

    const errs = extractErrors(data)
    const errProp: WithErr = {}
    if (errs.length) {
      errProp.err = errs[0].stack
    }

    const [first, ...rest] = data
    const msgProp: WithMsg = isPlainObject(first)
      ? {
        ...this.serialize(first),
        // @ts-ignore
        msg: format(...rest),
      }
      : {
        msg: format(first, ...rest),
      }

    const log: Log = {
      ...this.extraProps,
      ...errProp,
      ...msgProp,
      name: this.name,
      hostname,
      pid,
      level: levels[lvl],
      time: new Date(),
    }

    // eslint-disable-next-line no-console
    console[lvl](this.transform(log))
  }

  trace(...data: any[]) {
    this.write('trace', ...data)
  }

  debug(...data: any[]) {
    this.write('debug', ...data)
  }

  info(...data: any[]) {
    this.write('info', ...data)
  }

  warn(...data: any[]) {
    this.write('warn', ...data)
  }

  error(...data: any[]) {
    this.write('error', ...data)
  }

  child(props: {}) {
    return new Logger({
      ...this.options,
      extraProps: props,
    })
  }
}
