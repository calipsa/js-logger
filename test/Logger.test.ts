import { identity } from 'lodash'
import { omit } from 'lodash/fp'

import Logger from '../src/Logger'

const fakeConsole = {
  debug: identity,
  trace: identity,
  info: identity,
  warn: identity,
  error: identity,
}

const withoutDirtyProps = omit([
  'hostname',
  'pid',
  'time',
  'v',
])

const withoutErr = omit([
  'err',
])

const transform = (s: any) =>
  withoutDirtyProps(JSON.parse(s))

describe('Logger', () => {
  it('should output correct log', () => {
    const logger = new Logger({
      name: 'test-log',
      console: fakeConsole,
      minLevel: 'info',
      extraProps: {
        extra: 'xtra',
      },
    })
    const debug = logger.debug('DEBUG')
    expect(debug).toBeUndefined()

    const info = logger.info('Some message')
    expect(transform(info)).toMatchSnapshot()

    const infoWithReplacement = logger.info('Number %d', 42)
    expect(transform(infoWithReplacement)).toMatchSnapshot()

    const warning = logger.warn('Achtung!')
    expect(transform(warning)).toMatchSnapshot()

    const error = logger.error(new Error('ERROR!'))
    const errorObj = transform(error)
    expect(typeof errorObj.err).toBe('string')
    expect(typeof errorObj.msg).toBe('string')
    expect(withoutErr(errorObj)).toMatchSnapshot()
  })
})
