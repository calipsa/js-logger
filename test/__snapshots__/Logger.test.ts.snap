// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logger should output correct log 1`] = `
Object {
  "level": 30,
  "msg": "Some message",
  "name": "test-log",
  "severity": "INFO",
}
`;

exports[`Logger should output correct log 2`] = `
Object {
  "level": 30,
  "msg": "Number 42",
  "name": "test-log",
  "severity": "INFO",
}
`;

exports[`Logger should output correct log 3`] = `
Object {
  "level": 40,
  "msg": "Achtung!",
  "name": "test-log",
  "severity": "WARNING",
}
`;

exports[`Logger should output correct log 4`] = `
Object {
  "level": 40,
  "msg": "WARNING!",
  "name": "test-log",
  "severity": "WARNING",
}
`;

exports[`Logger should output correct log 5`] = `
Object {
  "level": 50,
  "msg": "ERROR!",
  "name": "test-log",
  "severity": "ERROR",
}
`;

exports[`Logger should output correct log 6`] = `
Object {
  "level": 60,
  "msg": "FATAL!!!",
  "name": "test-log",
  "severity": "FATAL",
}
`;

exports[`Logger should output correct log for a child 1`] = `
Object {
  "foo": "bar",
  "level": 30,
  "msg": "Some message for child",
  "name": "test-log",
  "severity": "INFO",
}
`;

exports[`Logger should output correct log for a child 2`] = `
Object {
  "extra": "xtra",
  "foo": "bar",
  "level": 30,
  "msg": "Some message for child",
  "name": "test-log",
  "severity": "INFO",
}
`;
